working/0546-patterns/patterns-feature-specification.md

```diff
@@ -183,12 +183,11 @@ Before introducing each pattern in detail, here is a summary with some examples:
 | [Logical-or][logicalOrPattern] | `subpattern1 \| subpattern2` |
 | [Logical-and][logicalAndPattern] | `subpattern1 & subpattern2` |
 | [Relational][relationalPattern] | `== expression`<br>`< expression` |
+| [Cast][castPattern] | `foo as String` |
 | [Null-check][nullCheckPattern] | `subpattern?` |
 | [Null-assert][nullAssertPattern] | `subpattern!` |
-| [Literal][literalPattern] | `123`, `null`, `'string'` |
-| [Constant][constantPattern] | `math.pi`, `SomeClass.constant` |
-| [Variable][variablePattern] | `foo`, `String str`, `_`, `int _` |
-| [Cast][castPattern] | `foo as String` |
+| [Constant][constantPattern] | `123`, `null`, `'string'`<br>`math.pi`, `SomeClass.constant`<br>`const Thing(1, 2)`, `const (1 + 2)` |
+| [Variable][variablePattern] | `foo`, `var bar`, `String str`, `_`, `int _` |
```

Comment:
I went back and forth on this several times, but ultimately I think having a single pattern that covers all of these was the simplest answer.

I could split out wildcards, but I still think it's important to support typed wildcards like `int _`. Those should behave like variable patterns in that they are only allowed in irrefutable contexts if the variable isn't a downcast (which makes them useless, but still), and do a type test when used in a refutable context. The only difference between them and variable patterns is that they don't bind. If I treated `type? '_'` as a separate pattern, I'd have to specify that ambiguity between those and typed variables is resolved to a wildcard (since `_` is a valid `identifier`). Also, `var _` should be non-binding to be consistent with *non*-pattern variable declarations, assuming we can pull off that change too. It just ended up being simpler to treat them as variable patterns.

Likewise, separating out identifiers into their own patterns doesn't buy much. If we use those to refer to named constants in refutable contexts, then the specification of them has to basically copy/paste the specification for constants but only when in a refutable context.

Reply:
If people can read it and understand which identifier means what where, then it works. Ship it.

It'll be Erik's problem to do the final formal specification anyway :smiling_imp: